enhancements/installer/connected-onprem-assisted-installer.md

```diff
@@ -0,0 +1,248 @@
+* a systemd unit file that runs all containers necessary to run the assisted installer
+
+Similar to downloading the current installer, the download will default to the
+latest available version of the assisted installer, RHCOS and OpenShift.
+
+### User Stories [optional]
+
+#### Story 1
+
+I want to install OpenShift while running the assisted installer and the entire
+installation workflow on infrastructure that I control. I do not want a third
+party cloud service to control the provisioning of my hosts from across the
+internet. I don't mind downloading content over the internet during the install
+workflow.
+
+#### Story 2
+
+I want to automate use of the assisted installer and expect reproducible
+results. Utilizing a cloud service does not provide sufficient control or
+guarantees of reproducibility.
+
+#### Story 3
+
+I want to test the assisted installer prior to production use and expect
+reproducible results. Utilizing a cloud service does not provide sufficient
+control or guarantees of reproducibility.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Single Use on Day 1 Only
+
+The on-premise assisted installer will run once per cluster. As a live ISO, it
+will not utilize any persistent storage. Upon completion of a cluster
+installation, the live ISO will be shut down and all state will vanish.
+
+#### Creating the Agent ISO
+
+The assisted installer generates a live ISO that can be booted on hosts and then
+used to provision them. It runs the Agent that communicates with the assisted
+installer service. The ISO needs to include:
+
+* a URL to where the assisted installer is running
+* a CA certificate so it can trust the assisted installer
+
+The assistant needs local access to a base RHCOS ISO in order to generate the
+Agent ISO. It will then run the following command, passing ignition either via
+`stdin` or as an additional argument.
+
+```
+coreos-installer iso embed -o /<output-path>/agent.iso /path/to/base/rhcos.iso
+```
+
+To facilitate local access, the base ISO image will be included inside the
+assisted installer's container image.
+
+#### S3
+
+The cloud implementation stores each generated ISO in AWS S3. This enables the
+service to scale its storage requirements as it generates many ISOs for the
+creation of many clusters.
+
+The on-premise implementation will only be used once to create a single
+cluster, so it does not require a scalable object store. Instead of running an
+S3-like service in the on-premise implementation, the assisted installer
+service will write the generated ISO to its filesystem and serve that file
+through its API directly from the filesystem.
+
+### Risks and Mitigations
+
+Communication between the Agent and the assisted installer must be protected
+with TLS. The Agent must be able to trust the assisted installer service so
+that it does not receive rogue provisioning instructions.
+
+## Design Details
+
+### Open Questions [optional]
+
+#### RHCOS ISO image delivery
```

Comment:
@cgwalters do you have any suggestions?

Reply:
This came up on internal chat - if you're booted into the Live ISO already, then you can just copy the bits from `/dev/sr0` (e.g.) right?

If you're talking about accessing the ISO content outside of the ISO itself, I think this is covered by https://github.com/openshift/enhancements/pull/201 right?

Reply:
We need to embed new ignition into a live ISO, and do that from within a container that's running in the live ISO.

I was aprehensive about using the block device, but maybe that's reasonable to mount it into the container. Could we then run `coreos-installer iso embed -o /<output-path>/agent.iso /dev/sr0`?

Reply:
Yep, this worked for me:
```
$ coreos-installer iso embed -f -c /path/to/new.ign /dev/sr0 -o /run/new.iso
```

Reply:
To be robust instead of `/dev/sr0` I'd look for `/dev/disk/by-label/*coreos*` - or alternatively `findmnt -n -o SOURCE /run/media/iso`.

Reply:
That's excellent. Thank you!